File: workspace_hub.py
"""Small owned queries and a server-clock focus timer; no AI requests."""
from contextlib import contextmanager
from datetime import datetime, time, timedelta, timezone
from functools import wraps
from uuid import UUID
from zoneinfo import ZoneInfo

from flask import jsonify, request


def focus_elapsed(row, now):
    elapsed = row['elapsed_seconds']
    if row['status'] == 'running' and row['last_started_at']:
        elapsed += max(0, int((now - row['last_started_at']).total_seconds()))
    return min(row['duration_seconds'], elapsed)


def search_pattern(value):
    query = value.strip()
    if not 2 <= len(query) <= 100:
        raise ValueError('Search using 2 to 100 characters.')
    return '%' + query.replace('\\', '\\\\').replace('%', '\\%').replace('_', '\\_') + '%'


def register(app, b):
    @contextmanager
    def db():
        conn = b['get_db']()
        try:
            with conn.cursor() as cur:
                yield conn, cur
        except Exception:
            conn.rollback()
            raise
        finally:
            conn.close()

    def private(fn):
        @wraps(fn)
        def wrapped(*args, **kwargs):
            uid = b['require_user_id']()
            if not uid:
                return jsonify(error='Please log in first.'), 401
            try:
                return fn(uid, *args, **kwargs)
            except (ValueError, TypeError):
                return jsonify(error='Check the search or focus session settings.'), 400
        return wrapped

    @app.get('/api/workspace/search')
    @private
    def search(uid):
        pattern = search_pattern(request.args.get('q', ''))
        with db() as (_, cur):
            cur.execute('''SELECT * FROM (
                (SELECT 'conversation' AS kind,c.id,c.title,
                    COALESCE((SELECT LEFT(m.content,160) FROM messages m WHERE m.conversation_id=c.id
                        AND m.user_id=%s AND m.content ILIKE %s ORDER BY m.id DESC LIMIT 1),'') AS excerpt,c.updated_at
                    FROM conversations c WHERE c.user_id=%s AND (c.title ILIKE %s OR EXISTS(
                        SELECT 1 FROM messages m WHERE m.conversation_id=c.id AND m.user_id=%s AND m.content ILIKE %s))
                    ORDER BY c.updated_at DESC LIMIT 6)
                UNION ALL
                (SELECT 'note',id,title,LEFT(content,160),updated_at FROM quick_notes
                    WHERE user_id=%s AND (title ILIKE %s OR content ILIKE %s) ORDER BY updated_at DESC LIMIT 6)
                UNION ALL
                (SELECT 'task',id,title,LEFT(details,160),updated_at FROM tasks
                    WHERE user_id=%s AND (title ILIKE %s OR details ILIKE %s) ORDER BY updated_at DESC LIMIT 6)
                ) found ORDER BY updated_at DESC LIMIT 18''',
                (uid, pattern, uid, pattern, uid, pattern, uid, pattern, pattern, uid, pattern, pattern))
            items = [{**dict(row), 'updated_at': row['updated_at'].isoformat()} for row in cur.fetchall()]
        return jsonify(items=items)

    @app.get('/api/workspace/today')
    @private
    def today(uid):
        now = datetime.now(timezone.utc)
        with db() as (_, cur):
            cur.execute('SELECT timezone FROM workspace_preferences WHERE user_id=%s', (uid,))
            pref = cur.fetchone()
            zone = ZoneInfo(pref['timezone'] if pref else 'UTC')
            tomorrow = datetime.combine(now.astimezone(zone).date() + timedelta(days=1), time.min, zone)
            cur.execute('''SELECT * FROM tasks WHERE user_id=%s AND completed=FALSE AND (due_at IS NULL OR due_at<%s)
                ORDER BY CASE WHEN due_at<%s THEN 0 ELSE 1 END,
                    CASE priority WHEN 'high' THEN 0 WHEN 'medium' THEN 1 ELSE 2 END,
                    due_at NULLS LAST,id LIMIT 3''', (uid, tomorrow, now))
            tasks = [b['task_to_dict'](row) for row in cur.fetchall()]
            cur.execute('SELECT COUNT(*) AS n FROM revision_items WHERE user_id=%s AND paused=FALSE AND next_review_at<=%s', (uid, now))
            due = cur.fetchone()['n']
            cur.execute('''SELECT id,title FROM conversations c WHERE user_id=%s AND is_archived=FALSE
                AND EXISTS(SELECT 1 FROM messages m WHERE m.conversation_id=c.id AND m.user_id=%s)
                ORDER BY updated_at DESC LIMIT 1''', (uid, uid))
            recent = cur.fetchone()
        return jsonify(tasks=tasks, revision_due=due, recent_chat=dict(recent) if recent else None)

    def settle(cur, row, now):
        if row and row['status'] == 'running' and focus_elapsed(row, now) >= row['duration_seconds']:
            cur.execute("""UPDATE focus_sessions SET status='completed',elapsed_seconds=duration_seconds,
                last_started_at=NULL,version=version+1 WHERE id=%s RETURNING *""", (row['id'],))
            return cur.fetchone()
        return row

    def serialized(row, now):
        if not row:
            return None
        return {'id': row['id'], 'task_id': row['task_id'], 'status': row['status'], 'version': row['version'],
                'duration_seconds': row['duration_seconds'], 'remaining_seconds': row['duration_seconds'] - focus_elapsed(row, now)}

    @app.route('/api/focus', methods=['GET', 'POST'])
    @private
    def focus(uid):
        now = datetime.now(timezone.utc)
        data = request.get_json(silent=True) or {}
        if not isinstance(data, dict):
            raise ValueError()
        with db() as (conn, cur):
            # Serialize starts and transitions per account, including across tabs.
            cur.execute('SELECT pg_advisory_xact_lock(82,%s)', (uid,))
            cur.execute('SELECT * FROM focus_sessions WHERE user_id=%s ORDER BY id DESC LIMIT 1 FOR UPDATE', (uid,))
            row = settle(cur, cur.fetchone(), now)
            if request.method == 'POST':
                client_id = str(UUID(str(data.get('client_id', ''))))
                minutes, task = data.get('minutes'), data.get('task_id')
                if type(minutes) is not int or minutes not in (25, 50) or (task is not None and type(task) is not int):
                    raise ValueError()
                cur.execute('SELECT * FROM focus_sessions WHERE user_id=%s AND client_id=%s', (uid, client_id))
                existing = cur.fetchone()
                if existing:
                    conn.commit()
                    return jsonify(session=serialized(existing, now))
                if row and row['status'] in ('running', 'paused'):
                    conn.commit()
                    return jsonify(error='Finish the current focus session first.', session=serialized(row, now)), 409
                if task is not None:
                    cur.execute('SELECT id FROM tasks WHERE id=%s AND user_id=%s AND completed=FALSE', (task, uid))
                    if not cur.fetchone():
                        return jsonify(error='Choose an open task from your planner.'), 404
                cur.execute('''INSERT INTO focus_sessions(user_id,task_id,client_id,duration_seconds,status,last_started_at)
                    VALUES(%s,%s,%s,%s,'running',%s) RETURNING *''', (uid, task, client_id, minutes * 60, now))
                row = cur.fetchone()
                cur.execute('''DELETE FROM focus_sessions WHERE user_id=%s AND status NOT IN ('running','paused')
                    AND id NOT IN (SELECT id FROM focus_sessions WHERE user_id=%s ORDER BY id DESC LIMIT 100)''', (uid, uid))
            conn.commit()
        return jsonify(session=serialized(row, now))

    @app.patch('/api/focus/<int:session_id>')
    @private
    def change_focus(uid, session_id):
        data = request.get_json(silent=True)
        if not isinstance(data, dict) or data.get('action') not in ('pause', 'resume', 'end') or type(data.get('version')) is not int:
            raise ValueError()
        now = datetime.now(timezone.utc)
        with db() as (conn, cur):
            cur.execute('SELECT pg_advisory_xact_lock(82,%s)', (uid,))
            cur.execute('SELECT * FROM focus_sessions WHERE id=%s AND user_id=%s FOR UPDATE', (session_id, uid))
            row = cur.fetchone()
            if not row:
                return jsonify(error='Focus session not found.'), 404
            if row['version'] != data['version']:
                return jsonify(error='This session changed in another tab. Refresh it before continuing.'), 409
            elapsed = focus_elapsed(row, now)
            action = data['action']
            if row['status'] in ('completed', 'cancelled'):
                return jsonify(session=serialized(row, now))
            if (action == 'pause' and row['status'] != 'running') or (action == 'resume' and row['status'] != 'paused'):
                return jsonify(error='Refresh the focus session before continuing.'), 409
            status = 'completed' if elapsed == row['duration_seconds'] else {'pause': 'paused', 'resume': 'running', 'end': 'cancelled'}[action]
            cur.execute('''UPDATE focus_sessions SET status=%s,elapsed_seconds=%s,last_started_at=%s,version=version+1
                WHERE id=%s RETURNING *''', (status, elapsed, now if status == 'running' else None, session_id))
            row = cur.fetchone()
            conn.commit()
        return jsonify(session=serialized(row, now))

File: workspace-hub.js
/* Search and focus are loaded only when opened. Today uses the existing data loader. */
(function(){
  'use strict';
  const state={api:null,user:null,notify:()=>{},openView:null,openTask:null};
  const t=value=>window.SaathiI18n?.t(value)||value;
  function node(tag,text,cls){const el=document.createElement(tag);if(text!=null)el.textContent=text;if(cls)el.className=cls;return el}
  function button(label,action,cls='secondary'){const el=node('button',t(label),cls);el.type='button';el.onclick=action;return el}
  function dialog(title){const d=node('dialog',null,'hub-dialog'),head=node('header'),h=node('h2',t(title));h.id='hub-title-'+crypto.randomUUID();d.setAttribute('aria-labelledby',h.id);head.append(h,button('Close',()=>d.close()));d.append(head);document.body.append(d);d.addEventListener('close',()=>d.remove());return d}
  function showError(box,error){box.textContent=t(error.message||'Something went wrong.');box.setAttribute('role','alert')}
  async function openItem(item){
    if(item.kind==='conversation'){location.href='/chat?conversation='+item.id;return}
    if(item.kind==='note'){await window.SaathiWorkspace.openNote(item.id);return}
    if(state.openTask){await state.openTask(item.id);return}
    location.href='/dashboard?task='+item.id+'#tasks';
  }
  function search(){
    if(!state.user)return;
    const d=dialog('Search your workspace'),input=node('input'),list=node('div',null,'hub-results'),message=node('p',t('Search chats, notes and tasks.'),'helper');
    input.type='search';input.maxLength=100;input.placeholder=t('Search chats, notes and tasks.');input.setAttribute('aria-label',t('Search your workspace'));message.setAttribute('role','status');
    d.append(input,message,list);let timeout,controller,sequence=0;
    async function run(){
      const query=input.value.trim(),turn=++sequence;controller?.abort();list.replaceChildren();
      if(query.length<2){message.textContent=t('Type at least two characters.');return}
      controller=new AbortController();message.textContent=t('Searching…');
      try{const data=await state.api('/api/workspace/search?q='+encodeURIComponent(query),{signal:controller.signal});if(turn!==sequence||!d.open)return;
        message.textContent=data.items.length?t('Results from your account.'):t('No matching items.');
        for(const item of data.items){const entry=node('button',null,'hub-result');entry.type='button';entry.append(node('small',t({conversation:'Conversation',note:'Note',task:'Task'}[item.kind])),node('strong',item.title),node('span',item.excerpt));entry.onclick=()=>{d.close();openItem(item).catch(error=>state.notify(t(error.message),'error'))};list.append(entry)}
      }catch(error){if(error.name!=='AbortError'&&turn===sequence)showError(message,error)}
    }
    input.oninput=()=>{clearTimeout(timeout);controller?.abort();sequence++;timeout=setTimeout(run,250)};
    input.onkeydown=event=>{if(event.key==='Enter'){event.preventDefault();clearTimeout(timeout);run()}};
    d.addEventListener('close',()=>{sequence++;clearTimeout(timeout);controller?.abort()});d.showModal();input.focus();
  }
  async function focus(taskId=null){
    const d=dialog('Focus session'),form=node('form'),label=node('label',t('Choose a task')),tasks=node('select'),lengthLabel=node('label',t('Session length')),length=node('select'),clock=node('p','25:00','hub-clock'),status=node('p',t('Loading…'),'helper'),actions=node('div',null,'care-actions');
    tasks.setAttribute('aria-label',t('Choose a task'));length.setAttribute('aria-label',t('Session length'));
    for(const minutes of [25,50]){const option=node('option',minutes+' '+t('minutes'));option.value=String(minutes);length.append(option)}
    label.append(tasks);lengthLabel.append(length);form.append(label,lengthLabel);d.append(form,clock,status,actions);d.showModal();
    let current=null,remaining=0,measured=Date.now(),interval,busy=false,settling=false,clientId=crypto.randomUUID();
    function tick(){const seconds=Math.max(0,remaining-(current?.status==='running'?Math.floor((Date.now()-measured)/1000):0));clock.textContent=String(Math.floor(seconds/60)).padStart(2,'0')+':'+String(seconds%60).padStart(2,'0');if(!seconds&&current?.status==='running'&&!settling){settling=true;refresh().catch(error=>showError(status,error))}}
    function render(session){current=session;remaining=session?.remaining_seconds??Number(length.value||25)*60;measured=Date.now();settling=false;const active=['running','paused'].includes(session?.status);tasks.disabled=length.disabled=active;
      status.textContent=t({running:'One task at a time. Your timer continues if this panel closes.',paused:'Paused. Continue when you are ready.',completed:'Session complete. Take a short break.',cancelled:'Session ended. Start again when you are ready.'}[session?.status]||'Choose one manageable task.');
      actions.replaceChildren();if(active){actions.append(button(session.status==='running'?'Pause':'Resume',()=>change(session.status==='running'?'pause':'resume')),button('End session',()=>change('end')))}else actions.append(button('Start focus',start,'primary'));
      actions.append(button('Refresh',()=>refresh().catch(error=>showError(status,error))));tick();clearInterval(interval);interval=setInterval(tick,1000);
    }
    async function refresh(){const data=await state.api('/api/focus');if(d.open)render(data.session)}
    async function operate(action){if(busy)return;busy=true;actions.querySelectorAll('button').forEach(el=>el.disabled=true);try{await action()}catch(error){showError(status,error)}finally{busy=false;actions.querySelectorAll('button').forEach(el=>el.disabled=false)}}
    function start(){return operate(async()=>{const data=await state.api('/api/focus',{method:'POST',body:JSON.stringify({minutes:Number(length.value),task_id:tasks.value?Number(tasks.value):null,client_id:clientId})});clientId=crypto.randomUUID();if(d.open)render(data.session)})}
    function change(action){return operate(async()=>{const data=await state.api('/api/focus/'+current.id,{method:'PATCH',body:JSON.stringify({action,version:current.version})});if(d.open)render(data.session)})}
    form.onsubmit=event=>{event.preventDefault();if(!current||!['running','paused'].includes(current.status))start()};length.onchange=()=>{remaining=Number(length.value)*60;tick()};
    const visible=()=>{if(!document.hidden&&d.open)refresh().catch(error=>showError(status,error))};document.addEventListener('visibilitychange',visible);d.addEventListener('close',()=>{clearInterval(interval);document.removeEventListener('visibilitychange',visible)});
    try{const [saved,planner]=await Promise.all([state.api('/api/focus'),state.api('/api/tasks')]);if(!d.open)return;const none=node('option',t('No task selected'));none.value='';tasks.append(none);for(const task of planner.tasks.filter(task=>!task.completed)){const option=node('option',task.title);option.value=task.id;option.setAttribute('data-user-content','');tasks.append(option)}tasks.value=String(saved.session?.task_id||taskId||'');render(saved.session)}catch(error){showError(status,error);actions.append(button('Retry',()=>{d.close();focus(taskId)}))}
  }
  function renderToday(data){
    const box=document.getElementById('todayTasks');if(!box)return;box.replaceChildren();
    if(!data.tasks.length)box.append(node('p',t('No tasks due. Add a small next step or leave room for rest.'),'helper'));
    for(const task of data.tasks){const row=node('div',null,'hub-task'),title=node('button',task.title,'hub-task-title');title.type='button';title.onclick=()=>openItem({kind:'task',id:task.id});const done=button('Complete',async()=>{done.disabled=true;try{await state.api('/api/tasks/'+task.id,{method:'PATCH',body:JSON.stringify({completed:true})});await state.refreshToday()}catch(error){done.disabled=false;state.notify(t(error.message),'error')}});row.append(title,button('Focus',()=>focus(task.id)),done);box.append(row)}
    const revisionLabel=node('span','Revision ready');revisionLabel.setAttribute('data-ui-text','');document.getElementById('todayRevision').replaceChildren(revisionLabel,node('span',': '+data.revision_due));
    const recent=document.getElementById('todayResume');recent.replaceChildren();if(data.recent_chat){const link=node('a',data.recent_chat.title,'secondary');link.href='/chat?conversation='+data.recent_chat.id;recent.append(link)}else recent.append(node('p',t('Your latest conversation will appear here.'),'helper'));
  }
  function connect(options){Object.assign(state,options);document.querySelectorAll('[data-workspace-search]').forEach(el=>el.onclick=search);document.querySelectorAll('[data-focus-session]').forEach(el=>el.onclick=()=>focus());}
  document.addEventListener('keydown',event=>{if((event.ctrlKey||event.metaKey)&&event.key.toLowerCase()==='k'&&!document.querySelector('dialog[open]')){event.preventDefault();search()}});
  window.SaathiHub={connect,search,focus,renderToday};
})();
